### DISCUSSION TITLE
Can a single package produce multiple versions?

### DISCUSSION THREAD

**Discussion by melMass** [AcademySoftwareFoundation/rez#1187] [Q&A] [answered]
Hi,

I'm trying to find a way to share `package code` **across DCC** versions or at least one for each major version.

I can't find the best workflow for that, here is what I tried:

- I created a custom `build_utils` python package in which I have a file per DCC, for instance `build_utils/templates/nuke.py`, in which I have all the package definition code besides the `version` attribute.
Then each version package have this import `from build_utils.templates.nuke import *`

This kind of works, but as I need to do a few things at build time, I can't use the `@early` decorator in my custom module, these need to be part of the `package` file itself.

- Then I tried another road, creating a few utility functions in my module: `get_vars(version, root)` which return all the env I need based on the version provided. And I wrap this in an @early function, here is my package for instance:

```python
from build_utils.templates.nuke import *

@early()
def get_vars(me):
    from build_utils.nuke import get_vars
    return get_vars(me["version"], this.root)

version = "13.0.2"
```

`this.root` here is not defined and the parameter keys does not have a root object, I'm trying to get the path to the source package because there I have a script I need to execute inside nuke but just at package build time.

Is there a way to do what I want in a cleaner way?

**Comment by davidlatwe** [comment]
The info you could get from rez in early stage is very limited. But if the package source path is all you need to get start, `os.getcwd()` should be enough. Since loading developer-package requires  `package.py` exists in current working directory.

**Comment by davidlatwe** [comment, marked as answer]
Additionally, to answer the title of this thread, there is one type of package that is not being documented in Wiki but actually does what you've asked for. It's called "combined-package", which looks like this (an example used in tests) :

https://github.com/nerdvegas/rez/blob/master/src/rez/data/tests/packages/py_packages/multi.py

That kind of packages aren't being structured with a package/version named directory plus a `package.py` in it, but only a python file exposed in packages dir path and named as `{pkg_name}.py`, e.g. `foo.py`, `nuke.py`.

**Reply by melMass** [reply]
Oh very nice! This is exactly what I was looking for. 
My main issue in fact with my method is that I need to use @early decorated functions because I want the values to be baked at build time.
I'm going to see how I can mix the two together

Thanks
